core/storage/question/gae_models.py

```diff
@@ -316,25 +316,70 @@ def get_question_skill_links_by_skill_ids(
+            random_id = unicode(start_cursor[-13:-1]) #pylint: disable=unicode-builtin
+            cursor = datastore_query.Cursor(urlsafe=start_cursor[:-13])
+            if direction == '1':
+                questions_list, next_cursor, more = cls.query(
+                    cls.skill_id.IN(skill_ids)
+                ).order(cls.question_id, cls.key).filter(
+                    cls.question_id > random_id
+                ).fetch_page(
+                    question_skill_count,
+                    start_cursor=cursor
+                )
+                if len(questions_list) < question_skill_count:
+                    direction = '0'
+                    question_list_extra, next_cursor, more = cls.query(
+                        cls.skill_id.IN(skill_ids)
+                    ).order(
+                        cls.question_id, cls.key
+                    ).filter(
+                        cls.question_id <= random_id
+                    ).fetch_page(
+                        question_skill_count - len(questions_list)
+                    )
+                    questions_list.extend(question_list_extra)
+            else:
+                questions_list, next_cursor, more = cls.query(
+                    cls.skill_id.IN(skill_ids)
+                ).order(cls.question_id, cls.key).filter(
+                    cls.question_id <= random_id
+                ).fetch_page(
+                    question_skill_count,
+                    start_cursor=cursor
+                )
         else:
-            question_skill_link_models, next_cursor, more = cls.query(
+            direction = '1'
+            random_id = utils.convert_to_hash(
+                python_utils.UNICODE(
+                    utils.get_random_int(base_models.RAND_RANGE)),
+                base_models.ID_LENGTH)
+            questions_list, next_cursor, more = cls.query(
                 cls.skill_id.IN(skill_ids)
-            ).order(-cls.last_updated, cls.key).fetch_page(
-                question_skill_count
-            )
+            ).order(cls.question_id, cls.key).filter(
+                cls.question_id > random_id
+            ).fetch_page(question_skill_count)
+            if len(questions_list) < question_skill_count:
+                direction = '0'
+                question_list_extra, next_cursor, more = cls.query(
+                    cls.skill_id.IN(skill_ids)
+                ).order(
+                    cls.question_id, cls.key
+                ).filter(
+                    cls.question_id <= random_id
+                ).fetch_page(
+                    question_skill_count - len(questions_list),
+                )
+                questions_list.extend(question_list_extra)
+
+        question_skill_link_models = questions_list
         next_cursor_str = (
             next_cursor.urlsafe() if (next_cursor and more) else None
         )
+        if next_cursor_str:
+            next_cursor_str = (next_cursor_str +
```

Comment:
Break after '('. See coding style guide (I mentioned this in a different PR review too).

Reply:
Done.